File: frontend/src/intel.js
// frontend/src/intel.js
// All Firestore operations — called from components and api.js
import {
  doc,
  getDoc,
  setDoc,
  updateDoc,
  arrayUnion,
  increment,
  serverTimestamp,
  collection,
  Timestamp,
} from 'firebase/firestore'
import { db } from './firebase'

// ── Helpers ───────────────────────────────────────────────────────────────

function ageDays(ts) {
  if (!ts) return 9999
  const ms = ts?.toDate ? ts.toDate().getTime() : new Date(ts).getTime()
  return Math.floor((Date.now() - ms) / 86400000)
}

export function freshnessLabel(age) {
  if (age <  1) return 'fresh'
  if (age <  7) return 'recent'
  if (age < 30) return 'aging'
  return 'stale'
}

function rangeKey(country, rirs) {
  return `${country.toUpperCase()}_${[...rirs].sort().join('_')}`
}

function cveKey(software, version) {
  const base = (software + (version ? '__' + version : ''))
    .toLowerCase().replace(/[\s/]/g, '_')
  return base.slice(0, 200)
}

// ══════════════════════════════════════════════════════════════════════════
// Range cache  (30-day TTL)
// ══════════════════════════════════════════════════════════════════════════

export async function getCachedRanges(country, rirs) {
  try {
    const key  = rangeKey(country, rirs)
    const snap = await getDoc(doc(db, 'range_cache', key))
    if (!snap.exists()) return null
    const data = snap.data()
    const age  = ageDays(data.cached_at)
    if (age >= 30) return null   // stale — caller fetches fresh
    return { ...data, from_cache: true, age_days: age, freshness: freshnessLabel(age) }
  } catch (e) {
    console.warn('[intel] range cache read:', e.message)
    return null
  }
}

export async function saveCachedRanges(country, rirs, result) {
  try {
    const key = rangeKey(country, rirs)
    await setDoc(doc(db, 'range_cache', key), {
      country:      result.country,
      country_name: result.country_name || '',
      rirs:         result.rirs || rirs,
      counts:       result.counts || {},
      total:        result.total  || result.ranges?.length || 0,
      ranges:       result.ranges || [],
      last_updated: result.last_updated || null,
      masscan_hint: result.masscan_hint  || null,
      cached_at:    serverTimestamp(),
    })
  } catch (e) {
    console.warn('[intel] range cache write:', e.message)
  }
}

// ══════════════════════════════════════════════════════════════════════════
// CVE cache  (7-day TTL)
// ══════════════════════════════════════════════════════════════════════════

export async function getCachedCves(software, version) {
  try {
    const key  = cveKey(software, version)
    const snap = await getDoc(doc(db, 'cve_cache', key))
    if (!snap.exists()) return null
    const data = snap.data()
    if (ageDays(data.cached_at) >= 7) return null
    return data.cves || []
  } catch (e) {
    console.warn('[intel] CVE cache read:', e.message)
    return null
  }
}

export async function saveCachedCves(software, version, cves) {
  try {
    const key = cveKey(software, version)
    await setDoc(doc(db, 'cve_cache', key), {
      software,
      version:   version || null,
      cves,
      cached_at: serverTimestamp(),
    })
  } catch (e) {
    console.warn('[intel] CVE cache write:', e.message)
  }
}

// ══════════════════════════════════════════════════════════════════════════
// Country intel  (crowdsourced scan results)
// ══════════════════════════════════════════════════════════════════════════

export async function getCountryIntel(country) {
  try {
    const snap = await getDoc(doc(db, 'country_intel', country.toUpperCase()))
    if (!snap.exists()) return { found: false, country: country.toUpperCase() }
    const data = snap.data()
    const age  = ageDays(data.last_scanned)
    return {
      ...data,
      found:     true,
      age_days:  age,
      freshness: freshnessLabel(age),
      stale:     age >= 30,
    }
  } catch (e) {
    console.warn('[intel] country intel read:', e.message)
    return { found: false, country: country.toUpperCase() }
  }
}

export async function contributeCountryIntel(country, payload) {
  try {
    const ref = doc(db, 'country_intel', country.toUpperCase())
    const snap = await getDoc(ref)

    const newHits = payload.hits || []

    // Merge hits — keep last 500
    const existingHits = snap.exists() ? (snap.data().hits || []) : []
    const mergedHits   = [...existingHits, ...newHits].slice(-500)

    // Top software frequency
    const swCount = {}
    for (const hit of mergedHits) {
      if (hit?.software) swCount[hit.software] = (swCount[hit.software] || 0) + 1
    }
    const topSoftware = Object.entries(swCount)
      .sort((a, b) => b[1] - a[1])
      .slice(0, 10)
      .map(([sw]) => sw)

    // CVE summary — unique, sorted by score desc
    const cveMap = {}
    for (const hit of mergedHits) {
      for (const cve of hit?.cves || []) {
        if (cve?.id && !cveMap[cve.id]) cveMap[cve.id] = cve
      }
    }
    const cveSummary = Object.values(cveMap)
      .sort((a, b) => parseFloat(b.score || 0) - parseFloat(a.score || 0))
      .slice(0, 20)

    const data = {
      country:      country.toUpperCase(),
      country_name: payload.country_name || '',
      last_scanned: serverTimestamp(),
      scan_count:   increment(1),
      stats: {
        probed: payload.probed || 0,
        open:   payload.open   || 0,
        vulns:  payload.vulns  || 0,
      },
      hits:         mergedHits,
      top_software: topSoftware,
      cve_summary:  cveSummary,
      ranges:       payload.ranges || [],
    }

    await setDoc(ref, data, { merge: true })
    console.log('[intel] contributed:', country.toUpperCase())
  } catch (e) {
    console.warn('[intel] contribute failed:', e.message)
  }
}

File: backend/intel.py
"""
backend/intel.py
Firestore-backed shared intelligence layer for XSEVERITY.
Handles country intel cache, CVE cache, and range cache.
"""

import os
from datetime import datetime, timedelta
from typing import Optional

# ── Firestore client (lazy init) ───────────────────────────────────────────
_db = None

def get_db():
    global _db
    if _db is None:
        from google.cloud import firestore
        # Set GOOGLE_APPLICATION_CREDENTIALS env var to your service account JSON
        # Or use Application Default Credentials if running on GCP
        _db = firestore.Client()
    return _db


# ══════════════════════════════════════════════════════════════════════════
# Freshness helpers
# ══════════════════════════════════════════════════════════════════════════

def _age_days(ts) -> int:
    """Return age in days from a Firestore timestamp or datetime."""
    if ts is None:
        return 9999
    if hasattr(ts, 'timestamp'):
        dt = datetime.fromtimestamp(ts.timestamp())
    else:
        dt = ts
    return (datetime.utcnow() - dt).days


def _freshness_label(age_days: int) -> str:
    if age_days < 1:
        return "fresh"
    if age_days < 7:
        return "recent"
    if age_days < 30:
        return "aging"
    return "stale"


# ══════════════════════════════════════════════════════════════════════════
# Range cache  (30-day TTL, shared across all users)
# ══════════════════════════════════════════════════════════════════════════

def _range_key(country: str, rirs: list[str]) -> str:
    return f"{country.upper()}_{'_'.join(sorted(r.upper() for r in rirs))}"


async def get_cached_ranges(country: str, rirs: list[str]) -> Optional[dict]:
    """
    Returns cached range data if fresh (< 30 days), else None.
    Caller should fall back to live RIR fetch.
    """
    try:
        db = get_db()
        key = _range_key(country, rirs)
        doc = db.collection("range_cache").document(key).get()
        if not doc.exists:
            return None
        data = doc.to_dict()
        age = _age_days(data.get("cached_at"))
        return {
            **data,
            "from_cache": True,
            "age_days": age,
            "freshness": _freshness_label(age),
            "stale": age >= 30,
        }
    except Exception as e:
        print(f"[intel] range cache read error: {e}")
        return None


async def save_cached_ranges(country: str, rirs: list[str], result: dict):
    """Save fresh range fetch result to Firestore."""
    try:
        db = get_db()
        key = _range_key(country, rirs)
        now = datetime.utcnow()
        db.collection("range_cache").document(key).set({
            "country": result["country"],
            "country_name": result.get("country_name", ""),
            "rirs": result.get("rirs", rirs),
            "counts": result.get("counts", {}),
            "total": result.get("total", len(result.get("ranges", []))),
            "ranges": result.get("ranges", []),
            "last_updated": result.get("last_updated"),
            "masscan_hint": result.get("masscan_hint"),
            "cached_at": now,
            "expires_at": now + timedelta(days=30),
        })
        print(f"[intel] range cache saved: {key}")
    except Exception as e:
        print(f"[intel] range cache write error: {e}")


# ══════════════════════════════════════════════════════════════════════════
# CVE cache  (7-day TTL, keyed by software string)
# ══════════════════════════════════════════════════════════════════════════

def _cve_key(software: str, version: Optional[str]) -> str:
    base = software.lower().replace(" ", "_").replace("/", "_")
    if version:
        base += f"__{version.lower().replace(' ', '_')}"
    return base[:200]  # Firestore doc ID limit


async def get_cached_cves(software: str, version: Optional[str]) -> Optional[list]:
    """Returns cached CVE list if < 7 days old, else None."""
    try:
        db = get_db()
        key = _cve_key(software, version)
        doc = db.collection("cve_cache").document(key).get()
        if not doc.exists:
            return None
        data = doc.to_dict()
        age = _age_days(data.get("cached_at"))
        if age >= 7:
            return None
        return data.get("cves", [])
    except Exception as e:
        print(f"[intel] CVE cache read error: {e}")
        return None


async def save_cached_cves(software: str, version: Optional[str], cves: list):
    """Save CVE lookup result to Firestore."""
    try:
        db = get_db()
        key = _cve_key(software, version)
        now = datetime.utcnow()
        db.collection("cve_cache").document(key).set({
            "software": software,
            "version": version,
            "cves": cves,
            "cached_at": now,
            "expires_at": now + timedelta(days=7),
        })
    except Exception as e:
        print(f"[intel] CVE cache write error: {e}")


# ══════════════════════════════════════════════════════════════════════════
# Country intel  (crowdsourced scan results per country)
# ══════════════════════════════════════════════════════════════════════════

async def get_country_intel(country: str) -> Optional[dict]:
    """Returns current community scan state for a country."""
    try:
        db = get_db()
        doc = db.collection("country_intel").document(country.upper()).get()
        if not doc.exists:
            return None
        data = doc.to_dict()
        age = _age_days(data.get("last_scanned"))
        return {
            **data,
            "found": True,
            "age_days": age,
            "freshness": _freshness_label(age),
            "stale": age >= 30,
        }
    except Exception as e:
        print(f"[intel] country intel read error: {e}")
        return None


async def contribute_country_intel(country: str, payload: dict):
    """
    Called when a scan completes. Merges new scan data into country intel.
    Appends hits (capped at 500), updates stats, increments scan count.
    """
    try:
        from google.cloud import firestore as _fs
        db = get_db()
        ref = db.collection("country_intel").document(country.upper())
        doc = ref.get()

        now = datetime.utcnow()
        new_hits = payload.get("hits", [])

        # Cap hits at 500 — keep most recent
        existing_hits = []
        if doc.exists:
            existing_hits = doc.to_dict().get("hits", [])
        merged_hits = (existing_hits + new_hits)[-500:]

        # Build top software list from all hits
        software_count: dict = {}
        for hit in merged_hits:
            sw = hit.get("software")
            if sw:
                software_count[sw] = software_count.get(sw, 0) + 1
        top_software = sorted(software_count, key=software_count.get, reverse=True)[:10]

        # Build CVE summary
        cve_map: dict = {}
        for hit in merged_hits:
            for cve in hit.get("cves", []):
                cid = cve.get("id")
                if cid and cid not in cve_map:
                    cve_map[cid] = cve
        cve_summary = sorted(
            cve_map.values(),
            key=lambda c: float(c.get("score", 0) or 0),
            reverse=True
        )[:20]

        update = {
            "country": country.upper(),
            "country_name": payload.get("country_name", ""),
            "last_scanned": now,
            "scan_count": _fs.Increment(1),
            "stats": {
                "probed": payload.get("probed", 0),
                "open": payload.get("open", 0),
                "vulns": payload.get("vulns", 0),
            },
            "hits": merged_hits,
            "top_software": top_software,
            "cve_summary": cve_summary,
        }

        ref.set(update, merge=True)
        print(f"[intel] country intel updated: {country.upper()}")
    except Exception as e:
        print(f"[intel] country intel write error: {e}")